specifyweb/workbench/upload/treerecord.py

```diff
@@ -19,33 +20,134 @@
+                raise ValueError(f"No treedefitems found for rank {rank_name} and base treedef {base_treedef_id}")
+            if treedefitems.count() > 1:
+                raise ValueError(f"Multiple treedefitems found for rank {rank_name} and base treedef {base_treedef_id}")
+            return treedefitems
+
+        def _get_treedef_id(
+            rank_name: str,
+            tree: str,
+            treedef_id: Optional[int],
+            base_treedef_id: Optional[int],
+        ) -> int:
+            filter_kwargs = _build_filter_kwargs(rank_name, treedef_id)
+            treedefitems = tree_model.objects.filter(**filter_kwargs)
+            
+            if not treedefitems.exists() and treedef_id is not None:
+                filter_kwargs = _build_filter_kwargs(rank_name)
+                treedefitems = tree_model.objects.filter(**filter_kwargs)
+
+            if treedefitems.count() > 1:
+                treedefitems = _filter_by_base_treedef_id(treedefitems, rank_name, base_treedef_id)
+
+            first_item = treedefitems.first()
+            if first_item is None:
+                raise ValueError(f"No treedefitems found for rank {rank_name}")
+
+            return first_item.treedef_id
+
+        treedef_id = _get_treedef_id(rank_name, tree, treedef_id, base_treedef_id)
+        return TreeRank(rank_name, treedef_id, tree_lower)
+
+    def check_rank(self) -> bool:
+        tree_model = get_tree_model(self.tree)
+        rank = tree_model.objects.filter(name=self.rank_name, treedef_id=self.treedef_id)
+        return rank.exists() and rank.count() == 1
+
+    def validate_rank(self) -> None:
+        if not self.check_rank():
+            raise ValueError(f"Invalid rank {self.rank_name} for treedef {self.treedef_id}")
+
+    def tree_rank_record(self) -> 'TreeRankRecord':
+        assert self.rank_name is not None, "Rank name is required"
+        assert self.treedef_id is not None, "Treedef ID is required"
+        assert self.tree is not None and self.tree.lower() in SPECIFY_TREES, "Tree is required"
+        return TreeRankRecord(self.rank_name, self.treedef_id)
+
+def get_tree_model(tree: str):
+    return getattr(models, tree.lower().title() + 'treedefitem')
+
+class TreeRankRecord(NamedTuple):
+    rank_name: str
+    treedef_id: int
+
+    def to_json(self) -> Dict:
+        return {
+            "rank": self.rank_name,
+            "treedefId": self.treedef_id,
+        }
+
+    def to_key(self) -> Tuple[str, int]:
+        return (self.rank_name, self.treedef_id)
+
+    def check_rank(self, tree: str) -> bool:
+        return TreeRank.create(self.rank_name, tree, self.treedef_id).check_rank()
+    
+    def validate_rank(self, tree) -> None:
+        TreeRank.create(self.rank_name, tree, self.treedef_id).validate_rank()
 
 class TreeRecord(NamedTuple):
     name: str
-    ranks: Dict[str, Dict[str, ColumnOptions]]
+    ranks: Dict[Union[str, TreeRankRecord], Dict[str, ColumnOptions]]
```

Comment:
I'll highly recommend getting rid of the union here. Make it consistent (doesn't matter which you choose - TreeRankRecord or str).

Reply:
Good suggestion, I was going for backwards compatibility here.  So you can still do it the old way if the rank name is unique in all trees, if not, it will just throw an error.